### PR TITLE
Add semantic behaviour tests for async loops (#395)

### LINKED ISSUE #395: Add semantic behaviour tests for async step execution loops
## Problem

Current tests for async loops may focus too heavily on generated code structure rather than semantic behaviour. Documentation drift was observed (fixed in `step_loop.rs:105-116`), indicating that semantic invariants need stronger test coverage independent of implementation details.

## Context

From PR #384 review: The async loop documentation required alignment with actual sync execution semantics, suggesting doc drift is an ongoing risk. Tests should enforce invariants (skip propagation, step ordering) independent of codegen structure.

## Proposed Solution

Add tests that assert semantic behaviour rather than code shape:

1. **Skip propagation tests**
   - Verify skipped steps halt execution
   - Assert subsequent steps do not run
   - Confirm skip messages are preserved
   - Validate skip position is recorded

2. **Step ordering tests**
   - Verify steps execute in declaration order
   - Confirm fixtures are available when needed
   - Validate example parameters appear after fixtures

3. **Error propagation tests**
   - Assert panics include feature/scenario metadata
   - Verify error messages contain step context
   - Confirm cleanup runs on failure

4. **Fixture lifecycle tests**
   - Validate RefCell borrows work across steps
   - Confirm fixtures inserted in one step are available in next
   - Verify fixture cleanup on scenario completion

## Implementation Strategy

Use property-based testing or behaviour-driven assertions that:
- Do not inspect generated token streams
- Assert runtime observable behaviour
- Are resilient to codegen refactoring
- Cover edge cases (empty scenarios, all-skipped, mixed async/sync)

## Acceptance Criteria

- [ ] Add skip propagation semantic tests
- [ ] Add step ordering semantic tests
- [ ] Add error propagation semantic tests
- [ ] Add fixture lifecycle semantic tests
- [ ] Ensure tests pass with current implementation
- [ ] Document testing strategy in contributor guide

## Effort

Small (S)

## References

- PR #384: https://github.com/leynos/rstest-bdd/pull/384
- Fixed doc drift: `crates/rstest-bdd-macros/src/codegen/scenario/runtime/generators/step_loop.rs:105-116`

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Introduces asynchronous semantic behaviour tests to validate runtime contracts of async loop execution, including skip propagation, step ordering, cross-step borrowing of fixtures, failure propagation, and cleanup guarantees.
- Adds a dedicated test module, shared test support utilities, a feature file, and a testing strategy document to distinguish semantic behaviour checks from structural macro tests.

### Tests
- New: crates/rstest-bdd/tests/async_semantic_behaviour.rs
  - Semantic, runtime-focused tests for async loop behavior (skip, order, fixture lifecycle, and error handling).
  - Includes scenarios that verify skip metadata, declaration order preservation, cross-step borrows with RefCell fixtures, and propagation of failures with context.
  - Uses serial and Tokio-based async tests to exercise concurrency boundaries.
- New: crates/rstest-bdd/tests/common/async_semantic_behaviour_support.rs
  - Shared support for semantic tests: event logging, cleanup probe tracking via Drop, feature path and scenario constants, and helpers to inspect bypass metadata when diagnostics are enabled.
  - Provides utilities to reset and inspect state across test scenarios.
- New: crates/rstest-bdd/tests/features/async_semantic_behaviour.feature
  - Feature file describing scenarios for async skip propagation, step ordering, returned fixtures, RefCell behavior across steps, and cleanup probes.
- New: docs/testing-strategy.md
  - Formal testing strategy distinguishing semantic behaviour tests from structural macro tests.
  - Outlines invariant-based testing patterns and recommended practices for runtime-observable assertions.

### Documentation & Guidelines
- Updated CONTRIBUTING.md with Testing Guidelines emphasizing semantic behaviour tests for scenario execution changes and keeping structural macro tests for code-generation details. (References/testing strategy now documented in docs/testing-strategy.md.)
- Added docs/developers-guide.md outlining internal test infrastructure and guidance for semantic tests.

## Why
- To ensure async loop semantics remain stable across refactors by validating runtime contracts rather than relying solely on macro-token generation details.
- To provide a clear separation between semantic behaviour tests (runtime invariants) and structural tests (compile-time/token-level checks).

## How to run
- Default: cargo test -p rstest-bdd
- Enable diagnostics bypass metadata tests: cargo test -p rstest-bdd --features diagnostics
- If running the whole workspace: cargo test --workspace --all-features

## Risk & Mitigations
- Tests rely on serialization of bypass metadata when diagnostics are enabled; ensure the diagnostics feature is active in CI for those checks.
- Some async tests use serial execution to avoid flakiness due to shared state; this is isolated to ensure deterministic outcomes.

## Test plan
- [x] Run semantic tests locally with and without the diagnostics feature to verify bypass metadata assertions.
- [x] Validate skip propagation, assertion of captured events, and that cleanup drops occur per scenario completion.
- [x] Confirm that RefCell-backed fixtures remain borrowable across async step boundaries.
- [x] Ensure error propagation includes step context, scenario name, and feature path in panics.

📎 **Task**: https://www.devboxer.com/task/903c7fc9-446f-44a2-b72d-ce0fa985099f

Closes #395

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[crates/rstest-bdd/tests/common/async_semantic_behaviour_support.rs](https://github.com/leynos/rstest-bdd/pull/475/files/9049fcf3964ec9f88e3e8dd3a3ca306db050c60b#diff-7380ba7c6c08a083e4322ea76334542fefe90699ac2f2f161829b8163f8d92f6)

Comment on file

~~~rust
pub(crate) fn assert_handler_failure_context(
    message: &str,
    expected_suffix: &str,
    scenario_name: &str,
    step_keyword: &str,
    step_text: &str,
    function_name: &str,
    handler_error: &str,
) {
    let normalized_message = normalize_message(message);
    let pattern = format!(
        r"Step failed at index .*?: .*?{step_keyword}.*?{step_text}.*?{function_name}.*?{handler_error}.*?\(feature: .*?{expected_suffix}, scenario: .*?{scenario_name}\)",
        step_keyword = regex::escape(step_keyword),
        step_text = regex::escape(step_text),
        function_name = regex::escape(function_name),
        handler_error = regex::escape(handler_error),
        expected_suffix = regex::escape(expected_suffix),
        scenario_name = regex::escape(scenario_name),
    );
    let matcher = Regex::new(&pattern)
        .unwrap_or_else(|error| panic!("handler-failure matcher should compile: {error}"));
    assert!(
        matcher.is_match(&normalized_message),
        "panic message should include the handler failure context: {message}",
    );
}
~~~

❌ New issue: [Excess Number of Function Arguments](https://codescene.io/projects/69391/delta?repo-id=1163913&review-id=475&biomarker=Excess+Number+of+Function+Arguments&filename=crates%2Frstest-bdd%2Ftests%2Fcommon%2Fasync_semantic_behaviour_support.rs&method=assert_handler_failure_context)
assert_handler_failure_context has 7 arguments, max arguments = 4

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[crates/rstest-bdd/tests/common/async_semantic_behaviour_support.rs](https://github.com/leynos/rstest-bdd/pull/475/files/9049fcf3964ec9f88e3e8dd3a3ca306db050c60b#diff-7380ba7c6c08a083e4322ea76334542fefe90699ac2f2f161829b8163f8d92f6)

Comment on file

~~~rust
//! Shared support for async semantic behaviour tests.
~~~

❌ New issue: [String Heavy Function Arguments](https://codescene.io/projects/69391/delta?repo-id=1163913&review-id=475&biomarker=String+Heavy+Function+Arguments&filename=crates%2Frstest-bdd%2Ftests%2Fcommon%2Fasync_semantic_behaviour_support.rs&method=)
In this module, 83.3% of all arguments to its 12 functions are strings. The threshold for string arguments is 39.0%

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

## Excess Number of Function Arguments
crates/rstest-bdd/tests/common/async_semantic_behaviour_support.rs: assert_handler_failure_context

### What lead to degradation?
assert_handler_failure_context has 7 arguments, max arguments = 4

### Why does this problem occur?
Functions with many arguments indicate either a) low cohesion where the function has too many responsibilities, or b) a missing abstraction that encapsulates those arguments.
The threshold for the Rust language is 4 function arguments.

### How to fix it?
Start by investigating the responsibilities of the function. Make sure it doesn't do too many things, in which case it should be split into smaller and more cohesive functions. Consider the refactoring [INTRODUCE PARAMETER OBJECT](https://refactoring.com/catalog/introduceParameterObject.html) to encapsulate arguments that refer to the same logical concept.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

## String Heavy Function Arguments
crates/rstest-bdd/tests/common/async_semantic_behaviour_support.rs: 

### What lead to degradation?
In this module, 83.3% of all arguments to its 12 functions are strings. The threshold for string arguments is 39.0%

### Why does this problem occur?
String is a generic type that fail to capture the constraints of the domain object it represents. In this module, 83 % of all function arguments are string types.

### How to fix it?
Heavy string usage indicates a missing domain language. Introduce data types that encapsulate the semantics. For example, a user_name is better represented as a constrained User type rather than a pure string, which could be anything.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

~~~
Please address the comments from this code review:

## Overall Comments
- The shared global state in `async_semantic_behaviour_support` (`EVENTS` and `CLEANUP_DROPS`) combined with `#[serial]` makes these tests sensitive to ordering; consider scoping state per test (e.g., via fixtures or newtypes) so that accidental concurrent execution in future tests can’t introduce subtle flakiness.
- The assertions in `error_propagation_includes_step_and_scenario_context` rely on multiple `message.contains(...)` checks for specific substrings; if there is any structured error metadata available (or a single, more constrained regex/message pattern), using that would make the test less brittle to wording changes while still validating the same semantics.
- Constants like `SKIP_SCENARIO_LINE: 4` tightly couple tests to the exact line layout of the feature file; if possible, derive or look up this information from the registry or parser instead of hard-coding line numbers to reduce maintenance when the feature file evolves.

## Individual Comments

### Comment 1
<location path="crates/rstest-bdd/tests/async_semantic_behaviour.rs" line_range="338-341" />
<code_context>
+
+#[test]
+#[serial]
+fn cleanup_probe_drops_after_failed_scenario_completion() {
+    reset_cleanup_drops();
+    let result = catch_unwind(semantic_cleanup_failure_scenario);
+    assert!(result.is_err(), "expected cleanup scenario to fail");
+    assert!(
+        cleanup_drops() > 0,
+        "fixtures should be dropped even when scenario execution fails",
+    );
</code_context>
<issue_to_address>
**suggestion (testing):** Tighten the cleanup-drop assertion for failed scenarios to match the success-path invariant

In `cleanup_probe_drops_after_successful_scenario_completion` you assert `cleanup_drops() == 1`, but here you only require `> 0`. If each scenario—successful or failed—is meant to trigger exactly one cleanup drop, this test should also assert `== 1` to catch multiple drops or repeated cleanup calls. If failures can legitimately cause multiple drops, consider documenting that in a comment to explain the weaker assertion.

```suggestion
    assert_eq!(
        cleanup_drops(),
        1,
        "fixtures should be dropped exactly once even when scenario execution fails",
    );
```
</issue_to_address>
~~~

File: crates/rstest-bdd/tests/async_semantic_behaviour.rs [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+    #[from(semantic_cleanup_probe)] _semantic_cleanup_probe: &CleanupProbe,
+) -> Result<(), &'static str> {
+    Err("cleanup probe failure")
+}
+
+#[scenario(
+    path = "tests/features/async_semantic_behaviour.feature",
+    name = "async skip propagation preserves metadata"
+)]
+async fn semantic_async_skip_scenario() {
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/tests/async_semantic_behaviour.rs:168]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/rstest-bdd/tests/async_semantic_behaviour.rs [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+#[ignore = "exercised by error_propagation_includes_step_and_scenario_context"]
+async fn semantic_async_error_scenario() {
+    panic!("scenario body should not execute after an earlier failure");
+}
+
+#[scenario(
+    path = "tests/features/async_semantic_behaviour.feature",
+    name = "cleanup probe completes successfully"
+)]
+fn semantic_cleanup_success_scenario(
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/tests/async_semantic_behaviour.rs:231]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/rstest-bdd/tests/common/async_semantic_behaviour_support.rs [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    };
+    let bypassed_steps = parsed
+        .get("bypassed_steps")
+        .and_then(Value::as_array)
+        .unwrap_or_else(|| panic!("registry dump should include bypassed_steps"));
+    assert!(
+        bypassed_steps.iter().any(|entry| {
+            entry.get("scenario_name") == Some(&Value::String(scenario_name.into()))
+                && entry.get("scenario_line").and_then(Value::as_u64)
+                    == Some(u64::from(scenario_line))
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/tests/common/async_semantic_behaviour_support.rs:253]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

### ❌ Failed checks (2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Developer Documentation | ⚠️ Warning | PR introduces undocumented internal test APIs (ScenarioRef, StepRef, etc.) in tests/common/async\_semantic\_behaviour\_support.rs without corresponding guidance in docs/testing-strategy.md or CONTRIBUTING.md. | Update docs/testing-strategy.md with a section documenting the test support infrastructure, including type/function explanations, code examples, and usage guidance for the newly introduced helpers. |
| Module-Level Documentation | ⚠️ Warning | Both new modules lack module-level docstrings: async\_semantic\_behaviour.rs and async\_semantic\_behaviour\_support.rs start directly with imports without explanatory documentation. | Add module-level docstrings using //! syntax to both async\_semantic\_behaviour.rs and async\_semantic\_behaviour\_support.rs before any imports, documenting their respective purposes and functionality. |

File: crates/rstest-bdd/tests/common/async_semantic_behaviour_support.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/tests/common/async_semantic_behaviour_support.rs:253]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: docs/testing-strategy.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+The review discussion that led to issue `#395` is a concrete reminder that
+documentation and runtime behaviour can drift independently of code shape.
+Semantic tests are the backstop that keeps those contracts aligned.
+
+## Test support infrastructure
+
+Async semantic behaviour tests share a support module at
+`tests/common/async_semantic_behaviour_support.rs`. The types and helpers below
+should be used instead of raw strings wherever assertions require structured
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/testing-strategy.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/rstest-bdd/tests/common/async_semantic_behaviour_support.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/tests/common/async_semantic_behaviour_support.rs:253]: @coderabbitai feedback from the engineer:

> The current diagnostics dump shape in [diagnostics.rs](https://github.com/leynos/rstest-bdd/blob/add-semantic-behavior-tests-sq2stk/crates/rstest-bdd/src/registry/diagnostics.rs "Open crates/rstest-bdd/src/registry/diagnostics.rs on GitHub (leynos/rstest-bdd@add-semantic-behavior-tests-sq2stk)") is:
> 
> - `line`: the step-definition source line
> - `scenario_line`: the feature-scenario line
> 
> `DumpedBypassedStep` explicitly serializes both fields, and the helper in [async\_semantic\_behaviour\_support.rs](https://github.com/leynos/rstest-bdd/blob/add-semantic-behavior-tests-sq2stk/crates/rstest-bdd/tests/common/async_semantic_behaviour_support.rs "Open crates/rstest-bdd/tests/common/async_semantic_behaviour_support.rs on GitHub (leynos/rstest-bdd@add-semantic-behavior-tests-sq2stk)") is correctly matching the scenario line against `entry.get("scenario_line")`.
> 
> I also re-ran the diagnostics test target unchanged:
> 
> - `cargo test -p rstest-bdd --features diagnostics --test async_semantic_behaviour`
> 
> That passed with `8 passed, 4 ignored`.
> 
> For the requested grep:
> 
> - `rg 'entry\.get\("scenario_line"\)' crates/`
> 
> Current hits are:
> 
> - [async\_semantic\_behaviour\_support.rs](https://github.com/leynos/rstest-bdd/blob/add-semantic-behavior-tests-sq2stk/crates/rstest-bdd/tests/common/async_semantic_behaviour_support.rs "Open crates/rstest-bdd/tests/common/async_semantic_behaviour_support.rs on GitHub (leynos/rstest-bdd@add-semantic-behavior-tests-sq2stk)")
> - [dump\_registry.rs](https://github.com/leynos/rstest-bdd/blob/add-semantic-behavior-tests-sq2stk/crates/rstest-bdd/tests/dump_registry.rs "Open crates/rstest-bdd/tests/dump_registry.rs on GitHub (leynos/rstest-bdd@add-semantic-behavior-tests-sq2stk)")
> 
> Changing the helper to `entry.get("line")` would compare the scenario line against the step-definition line and would be incorrect for the current payload. No edits were made.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

### ❌ Failed checks (2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Developer Documentation | ⚠️ Warning | Pull request introduces substantial internal test APIs without required documentation in docs/developers-guide.md. | Create docs/developers-guide.md and consolidate all developer-facing documentation about internal test infrastructure there, ensuring all internal test API structs and helper functions are documented. |
| Module-Level Documentation | ⚠️ Warning | Multiple public and pub(crate) functions and structures in new test modules lack proper documentation, including nine public functions and two public structs without docstrings. | Add comprehensive doc comments to all public and pub(crate) items using /// syntax, documenting purpose, parameters, return values, panic conditions, and thread-local state handling. |

**Comment by leynos** [comment]
@coderabbitai pause

File: crates/rstest-bdd/tests/common/async_semantic_behaviour_support.rs [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+/// matching all fields of `query`.
+///
+/// Dumps the registry via `rstest_bdd::dump_registry`, parses the JSON, and
+/// searches `bypassed_steps` for an entry matching `scenario_name`,
+/// `scenario_line`, `step_pattern`, and a `reason` substring.
+///
+/// # Panics
+///
+/// Panics if the dump fails, the JSON is invalid, or no matching entry is found.
+pub(crate) fn assert_bypassed_step_recorded(query: BypassedStepQuery<'_>) {
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/tests/common/async_semantic_behaviour_support.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

### ❌ Failed checks (2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Developer Documentation | ⚠️ Warning | Documentation contains three unresolved style violations: heading not in sentence case, two instances of non-Oxford spelling, and misplaced cfg attribute. | Change 'Developer Guide' to 'Developer guide' on line 1; replace 'Normalises' with 'Normalizes' on lines 84–85 and 133; move cfg attribute after doc comment at line 220. |
| Module-Level Documentation | ⚠️ Warning | Struct fields lack doc comments and attribute placement violates conventions; spelling uses -ise instead of -ize. | Add doc comments for all pub(crate) struct fields, reposition #\[cfg\] attribute after doc comments, and replace 'Normalises' with 'Normalizes'. |
